[rb] Add support for beta chrome

aguspe opened the pull request:
### **User description**
### Motivation and Context
The purpose of this PR is to add the ability for ruby tests to run using chrome-beta to make it easier to debug

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Added support for running tests with Chrome Beta.

- Introduced `chrome_beta` method to configure Chrome Beta driver.

- Updated Bazel build files to include Chrome Beta configuration.

- Enhanced test environment to handle Chrome Beta driver setup.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_environment.rb</strong><dd><code>Add Chrome Beta driver support in test environment</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb

<li>Added <code>chrome_beta</code> method to configure Chrome Beta driver.<br> <li> Modified <code>create_driver!</code> to handle Chrome Beta driver.<br> <li> Adjusted driver method selection logic for Chrome Beta.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15417/files#diff-884d159fa58dcbf2cdfec0003a9a72bf2a95fa661f3d2787ec801a3669521b6a">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>tests.bzl</strong><dd><code>Add Chrome Beta configuration to Bazel tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/tests.bzl

<li>Added Chrome Beta configuration to Bazel test definitions.<br> <li> Included environment variables for Chrome Beta setup.<br> <li> Configured Chrome Beta dependencies and data.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15417/files#diff-0b035bed60fc62ad1db7c3548fa4df1aa1cd0ebe7b3c50cf5643a34929d8cb68">+23/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>BUILD.bazel</strong><dd><code>Include Chrome Beta in supported browsers list</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rb/spec/integration/BUILD.bazel

- Added `chrome-beta` to the list of supported browsers.


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15417/files#diff-2cb9bf7ae4b1c7648e110a9b7608c2883e196b4f63566711ce0340ea803ad63f">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

titusfortner commented on scripts/pinned_browsers.py:
```diff
@@ -26,12 +26,12 @@ def calculate_hash(url):
-    channel = args.chrome_channel
-
+def get_chrome_milestone(channel=None):
+    if channel is None:
```
Somewhere in the refactoring of this method something was removed. None of this code is needed, just set the default to 'Stable' and delete lines 30-34 here.

titusfortner commented on scripts/pinned_browsers.py:
```diff
@@ -48,18 +48,15 @@ def get_chrome_milestone():
-        content
-        + """    http_archive(
-        name = "linux_chromedriver",
+    content = content + """    http_archive(
```
There's a reason everything is formatted the way it is. I don't remember that reason, but I made a point of doing it that way: 163167e30f4168edbd7f4435e98dca87346f2481

titusfortner commented on rb/spec/tests.bzl:
```diff
@@ -32,6 +33,29 @@ BROWSERS = {
+        "target_compatible_with": [],
+        "env": {
+            "WD_REMOTE_BROWSER": "chrome",
+            "WD_SPEC_DRIVER": "chrome-beta",
```
we don't want to make these values different, and whatever we do here, we should also do to firefox. I think it might make sense to do `-beta` for all of them and sort it out in test_environment.

titusfortner commented on rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb:
```diff
@@ -313,6 +315,15 @@ def random_port
           sock.close
         end
+
+        def select_driver(opts)
```
I think we can just alias chrome_beta_driver and chrome_driver, and just make the changes to browser version in #chrome_beta_options and #firefox_beta_options

titusfortner commented on rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb:
```diff
@@ -195,7 +196,8 @@ def build_options(**opts)
           else
-            WebDriver::Options.send(browser, **opts)
+            parsed_browser = browser == :chrome_beta ? :chrome : browser
+            WebDriver::Options.send(parsed_browser, **opts)
```
if you create the options methods you don't have to change the default

titusfortner commented on rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb:
```diff
@@ -166,11 +166,12 @@ def root
                        send(method, listener: listener, options: build_options(**opts))
                      else
-                       WebDriver::Driver.for(driver, listener: listener, options: build_options(**opts))
+                       WebDriver::Driver.for(method, listener: listener, options: build_options(**opts))
```
if we create the driver methods, we don't have to modify the default

titusfortner reviewed (changes requested):
I like what you've done with most of this.

I keep going back and forth on if we want to pass chrome_beta or if we want to have a `WD_BROWSER_VERSION` environment variable for "stable" and "beta". I did the review thinking the former, and now I'm thinking the latter would be better. Mostly I care about adding: `guards.add_condition(:version, GlobalTestEnv.version)` which would allow us to do guards of `except: {browser: :chrome, version: :stable}`

I wonder if we should start with adjusting the firefox beta Ruby implementation to do what we want then we can copy it with the chrome.

aguspe commented:
> I like what you've done with most of this.
> 
> I keep going back and forth on if we want to pass chrome_beta or if we want to have a `WD_BROWSER_VERSION` environment variable for "stable" and "beta". I did the review thinking the former, and now I'm thinking the latter would be better. Mostly I care about adding: `guards.add_condition(:version, GlobalTestEnv.version)` which would allow us to do guards of `except: {browser: :chrome, version: :stable}`
> 
> I wonder if we should start with adjusting the firefox beta Ruby implementation to do what we want then we can copy it with the chrome.

Thank you @titusfortner for the review! appreciated, I'm trying to addressed the comments today and tomorrow, I will be resolving the remaining ones until it works and ping you for a review again :)

titusfortner commented:
I did another review of what Alex did for Firefox, and we can just copy exactly what Firefox does for chrome. There are good reasons we don't want to implement browser version environment variable right now

titusfortner commented:
You need to update the browser versions again:

```
bazel run //scripts:pinned_browsers
```

titusfortner commented:
@cgoldberg do we lint the python files in the scripts directory? If not, is it an easy add?

cgoldberg commented:
We only lint stuff under /py  ... I can add scripts

titusfortner commented:
@aguspe I think this is running into the same issue we hit with Python & JS where tests are being skipped that aren't obvious. Let's me remove them from the file and try to add guards with reasons. The JS test is passing in trunk, now, so I'll merge that one.

p0deje commented on common/repositories.bzl:
```diff
@@ -270,6 +268,84 @@ package(default_visibility = ["//visibility:public"])
+
+    http_archive(
+        name = "linux_beta_chrome",
+        url = "https://storage.googleapis.com/chrome-for-testing-public/137.0.7151.6/linux64/chrome-linux64.zip",
```
This seems to be outdated already as the current release is 137.0.7151.40
